haystack/document_stores/pinecone.py

```diff
@@ -865,39 +905,50 @@ def _get_all_document_ids(
-                    # Move these IDs to new namespace
-                    self._move_documents_by_id_namespace(
-                        ids=vector_id_matrix,
-                        index=index,
-                        source_namespace=namespace,
-                        target_namespace=target_namespace,
-                        batch_size=batch_size,
-                    )
-                    progress_bar.set_description_str("Retrieved IDs")
-                    progress_bar.update(len(set(vector_id_matrix)))
+            if pinecone.describe_index(index).pod_type == PINECONE_STARTER_POD:
+                # Due to missing support for Namespace in Starter Pinecone index type, retrieve up to 10000 vectors
+                logger.warning(
+                    "Current index type 'Starter' doesn't support 'Namespace' feature. "
+                    "Limit for fetching documents in 'Starter' index type is %s.",
+                    self.top_k_limit,
+                )
+                all_ids = self._get_ids(
+                    index=index, filters=filters, type_metadata=type_metadata, batch_size=self.top_k_limit
+                )
+            else:
+                # If we don't have all IDs, we must query and extract IDs from the original namespace, then move the retrieved embeddings
+                # to a temporary namespace and query again for new items. We repeat this process until all embeddings
+                # have been retrieved.
+                target_namespace = f"{namespace}-copy" if namespace is not None else "copy"
+                all_ids: Set[str] = set()  # type: ignore
+                vector_id_matrix = ["dummy-id"]
+                with tqdm(
+                    total=document_count, disable=not self.progress_bar, position=0, unit=" ids", desc="Retrieving IDs"
+                ) as progress_bar:
+                    while vector_id_matrix:
+                        # Retrieve IDs from Pinecone
+                        vector_id_matrix = self._get_ids(
+                            index=index,
+                            namespace=namespace,
+                            filters=filters,
+                            type_metadata=type_metadata,
+                            batch_size=batch_size,
+                        )
+                        # Save IDs
+                        all_ids = all_ids.union(set(vector_id_matrix))  # type: ignore
+                        # Move these IDs to new namespace
+                        self._move_documents_by_id_namespace(
+                            ids=vector_id_matrix,
+                            index=index,
+                            source_namespace=namespace,
+                            target_namespace=target_namespace,
+                            batch_size=batch_size,
+                        )
+                        progress_bar.set_description_str("Retrieved IDs")
+                        progress_bar.update(len(set(vector_id_matrix)))
```

Comment:
We can avoid recreating the `set` multiple times this way.
```suggestion
                        unique_ids = set(vector_id_matrix)
                        all_ids = all_ids.union(unique_ids)  # type: ignore
                        # Move these IDs to new namespace
                        self._move_documents_by_id_namespace(
                            ids=vector_id_matrix,
                            index=index,
                            source_namespace=namespace,
                            target_namespace=target_namespace,
                            batch_size=batch_size,
                        )
                        progress_bar.set_description_str("Retrieved IDs")
                        progress_bar.update(len(unique_ids))
```